data_model/src/domain.rs

```diff
@@ -46,75 +49,229 @@ impl From<GenesisDomain> for Domain {
+        self.accounts.values()
+    }
+
+    /// Return `true` if the `Domain` contains [`Account`]
+    #[inline]
+    pub fn contains_account(&self, account_id: &<Account as Identifiable>::Id) -> bool {
+        self.accounts.contains_key(account_id)
+    }
+
+    /// Get an iterator over asset definitions of the `Domain`
+    #[inline]
+    pub fn asset_definitions(&self) -> impl ExactSizeIterator<Item = &AssetDefinitionEntry> {
+        self.asset_definitions.values()
+    }
+}
+
+#[cfg(feature = "mutable_api")]
+impl Domain {
+    /// Return a mutable reference to the [`Account`] corresponding to the account id.
+    #[inline]
+    pub fn account_mut(
+        &mut self,
+        account_id: &<Account as Identifiable>::Id,
+    ) -> Option<&mut Account> {
+        self.accounts.get_mut(account_id)
+    }
+
+    /// Add [`Account`] into the [`Domain`] returning previous account stored under the same id
+    #[inline]
+    pub fn add_account(&mut self, account: impl Into<Account>) -> Option<Account> {
+        let account = account.into();
+        self.accounts.insert(account.id().clone(), account)
+    }
+
+    /// Remove account from the [`Domain`] and return it
+    #[inline]
+    pub fn remove_account(
+        &mut self,
+        account_id: &<Account as Identifiable>::Id,
+    ) -> Option<Account> {
+        self.accounts.remove(account_id)
+    }
+
+    /// Get a mutable iterator over accounts of the domain
+    #[inline]
+    pub fn accounts_mut(&mut self) -> impl ExactSizeIterator<Item = &mut Account> {
+        self.accounts.values_mut()
+    }
+
+    /// Get a mutable iterator over asset definitions of the [`Domain`]
+    #[inline]
+    pub fn asset_definition_mut(
+        &mut self,
+        asset_definition_id: &<AssetDefinition as Identifiable>::Id,
+    ) -> Option<&mut AssetDefinitionEntry> {
+        self.asset_definitions.get_mut(asset_definition_id)
+    }
+
+    /// Add asset definition into the [`Domain`] returning previous asset definition stored under
+    /// the same id
+    #[inline]
+    pub fn add_asset_definition(
+        &mut self,
+        asset_definition: impl Into<AssetDefinition>,
+        registered_by: <Account as Identifiable>::Id,
+    ) -> Option<AssetDefinitionEntry> {
+        let asset_definition = AssetDefinitionEntry::new(asset_definition.into(), registered_by);
+
+        self.asset_definitions
+            .insert(asset_definition.definition().id().clone(), asset_definition)
+    }
+
+    /// Remove asset definition from the [`Domain`] and return it
+    #[inline]
+    pub fn remove_asset_definition(
+        &mut self,
+        asset_definition_id: &<AssetDefinition as Identifiable>::Id,
+    ) -> Option<AssetDefinitionEntry> {
+        self.asset_definitions.remove(asset_definition_id)
+    }
+}
+
+impl From<<Self as Identifiable>::Constructor> for Domain {
+    fn from(source: <Self as Identifiable>::Constructor) -> Self {
```

Comment:
this is the only way to construct new Domain because Domain::new returns NewDomain. The same goes for other Identifiable types. This way we are protected from ill defined structures

Reply:
I can't see, why things like `Domain` can be ill-defined, but things like `Trigger` can not. Can you explain, please, what's the difference here?

Reply:
For `Domain` it's not valid for a user to define accounts ahead of time. User can add accounts only through separate instructions. Maybe some field of triggers should follow the same reasoning?

Reply:
Okay, I got it. Then `Action::technical_account` from `Trigger::action` field must already exist before trigger registration. Is it suitable for what you are doing?